jgrapht-core/src/main/java/org/jgrapht/graph/concurrent/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1416 @@
+ * concurrent read/write use case is add-only.
+ * <br>
+ * eg: If threadA tries to get all edges touching a certain vertex after threadB removes the vertex,
+ * the algorithm will be interrupted by {@link IllegalArgumentException}.
+ * </p>
+ * <pre>
+ *      Thread threadA = new Thread(() -&gt; {
+ *          Set vertices = graph.vertexSet();
+ *          for (Object v : vertices) {
+ *              // {@link IllegalArgumentException} may be thrown since other threads may have removed the vertex.
+ *              Set edges = graph.edgesOf(v);
+ *              doOtherThings();
+ *          }
+ *      });
+ *      Thread threadB = new Thread(() -&gt; {
+ *          Set vertices = graph.vertexSet();
+ *          for (Object v : vertices) {
+ *              if (someConditions)
+ *                  graph.removeVertex(v);
+ *          }
+ *      });
+ * </pre>
+ *
+ * <p>
+ * The created Graph's hashCode is equal to the backing set's hashCode. And the created Graph is equal
+ * to another Graph if they are the same Graph or the backing Graph is equal to the other Graph.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author CHEN Kui
+ * @since Feb 23, 2018
+ */
+public class AsSynchronizedGraph<V, E>
+    extends GraphDelegator<V, E>
+    implements Graph<V, E>, Serializable
+{
+    private static final long serialVersionUID = 5144561442831050752L;
+
+    private final ReentrantReadWriteLock readWriteLock;
+
+    // A set encapsulating backing vertexSet.
+    private transient CopyOnDemandSet<V> allVerticesSet;
+
+    // A set encapsulating backing edgeSet.
+    private transient CopyOnDemandSet<E> allEdgesSet;
+
+    private CacheStrategy<V, E> cacheStrategy;
+
+    /**
+     * Constructor for AsSynchronizedGraph with strategy of not caching the copies for
+     * <code>edgesOf</code>, <code>incomingEdgesOf</code> and <code>outgoingEdgesOf</code> methods
+     * and non-fair mode for thread-access.
+     *
+     * @param g the backing graph (the delegate)
+     */
+    public AsSynchronizedGraph(Graph<V, E> g)
+    {
+        this(g, false, false);
+    }
+
+    /**
+     * Constructor for AsSynchronizedGraph with specified properties.
+     *
+     * @param g the backing graph (the delegate)
+     * @param cacheEnable a flag describing whether a cache will be used
+     * @param fair a flag describing whether fair model will be used
+     */
+    private AsSynchronizedGraph(Graph<V, E> g, boolean cacheEnable, boolean fair)
+    {
+        super(g);
+        readWriteLock = new ReentrantReadWriteLock(fair);
+        if (cacheEnable)
+            cacheStrategy = new CacheAccess();
+        else
+            cacheStrategy = new NoCache();
+        allEdgesSet = new CopyOnDemandSet<>(super.edgeSet(), readWriteLock);
+        allVerticesSet = new CopyOnDemandSet<>(super.vertexSet(), readWriteLock);
+        for (V v : allVerticesSet) {
```

Comment:
Here and in addVertex, there should be a comment explaining precisely why these calls are needed, since it won't be obvious to most people reading the code.

Reply:
Ok

Reply:
Also, I created https://github.com/jgrapht/jgrapht/issues/520 so that maybe we can get the underlying issue resolved.

Reply:
Thanks, if any good solution, I'll help to implement.